### DISCUSSION TITLE
CSLA calls operations of base classes

### DISCUSSION THREAD

**Discussion by StefanOssendorf** [MarimerLLC/csla#4074] [Questions]
I'm wondering if this is a bug or a feature.
```csharp
public abstract class FooBase : BusinessBase<FooBase> {
  [Create]
  private void CreateBase() {
  }
}

public class Foo : FooBase {
}


IDataPortal<Foo> x;
var foo = await x.CreateAsync(); // This calls CreateBase()
```

Should the `CreateBase()` method be considered the `Create` operation of `Foo`? For me that looks like invoking a "constructor" only from the base class but _not_ from the actual class which could create a lot of trouble.

Edit: The same holds for non abstract classes, too. So "removing" a CSLA operation attribute is not necessarily an option.

**Comment by rockfordlhotka** [comment]
I think you are suggesting that the data portal should _not_ honor data portal operation attributes inherited from base types?

That does seem like it should make for more understandable code.

_Originally_, before the modern attribute scheme, the data portal relied on overrides of base methods like `DataPortal_Create`, and the base classes provided default implementations of (specifically) create. That is no longer valid, and those base class defaults are now gone.

My first instinct is to say that this is a bug - unintended behavior.

**Reply by rockfordlhotka** [reply]
I guess the question is whether there's value in someone creating a polymorphic set of business types, where all the types can have their data access in a base class (implying the same properties), and then different methods/rules/behaviors in a set of different subclasses.

From an OO perspective this is entirely valid. From a pragmatic perspective, it seems pretty niche. And in such a case, someone could have those various subclasses implement data portal operation methods that delegate to the base class, so the barrier to implementation seems quite low.

**Reply by StefanOssendorf** [reply]
tbh I don't know whats better or not. I stumbled across that in some code at work and wondered if it's working with inheritance or not. And during implementing to support this with my source gen I was wondering if this is intended or not.
From an OO perspective I don't think this is valid. We have private methods which should not be able to be called from the outside without going through an public API. I personally think of CSLA operation methods as "constructors" to my type to get an initialized instance. With the example from above we are breaking the "rule" that any initialization is going through the actual class up to the base classes. We are calling directly the method from the base class without any way to alter that behavior which would be always possible with a real ctor.

My question would be: What's the path with less unintended errors? 
(Furthermore any change to that behavior is a breaking change.)
My first gut feeling says it's weird/unexpected behavior and hence unintended.

**Reply by rockfordlhotka** [reply]
> What's the path with less unintended errors?

Also, what is the option that has a reasonable workaround?

Using the base class attributes has, afaik, no workaround. You are stuck with those overloads even if they aren't valid.

Using only the "leaf" class attributes has a workaround, in that you can implement attributed methods and delegate to the base method if necessary and appropriate.

**Reply by StefanOssendorf** [reply]
We could get fancy (and complicated) with:
* `private` csla method defined in a base class is _not_ resolved
* `protected` csla method defined in a base class is _not_ resolved
* `protected virtual` csla method defined in a base _is_ resolved

But I don't think this won't be overly confusing if implemented and used 😆.

So only the csla methods of the type in question (leaf is not entirely correct here) should be resolved and methods defined by base class(es) are ignored.
That would mean we have a bug/unintended behavior we should fix?

Some examples which - with the assumption from above - should not work any longer are:
```csharp
public abstract class FooBase : BusinessBase<FooBase> {
  [Create]
  private void CreateBase() {
  }
}

public class Foo : FooBase {
}
```
Foo can't be created due to missing csla method.

```csharp
public class BarBase : BusinessBase<BarBase> {
  [Create]
  private void CreateBarBase() {
  }
}

public class Bar : BarBase {
}
```
Bar can't be created _but_ BarBase can be created with an `IDataPortal<BarBase>`.

**Reply by StefanOssendorf** [reply]
@rockfordlhotka 
Just to sum this up.
This is unintended and we'll create an issue to get this changed in the future?

**Reply by rockfordlhotka** [reply]
@StefanOssendorf yes, would you please create an issue?

**Reply by StefanOssendorf** [reply]
Yes. I'll create an isseu for that.

**Reply by StefanOssendorf** [reply]
#4090 
Created :)

**Reply by michaelcsikos** [reply]
I would still expect `protected` to resolve without `virtual`. It's visible and can be called from an inheriting class.
